### ISSUE TITLE
BUG: Inconsistent Matplot layouts

### ISSUE THREAD

**Issue by dsgibbons** [shap/shap#3213]
### Issue Description

While removing the Boston dataset in #3200, it was observed that it is difficult to specify the layout of the resulting waterfall plot correctly. Please take a look at the minimal reproducible example below. Three variations have been tried so far. `plt.tight_layout()` would be my natural choice, but it leads to an uncentred plot in fullscreen mode. `plt.rcParams['figure.constrained_layout.use'] = True`  seems to always look good, but @thatlittleboy raised some concerns [here](https://github.com/shap/shap/pull/3200#discussion_r1302016585).

```python
import xgboost
import shap
import matplotlib.pyplot as plt

X, y = shap.datasets.california()
model = xgboost.XGBRegressor().fit(X, y)
explainer = shap.Explainer(model)
shap_values = explainer(X)
fig = shap.plots.waterfall(shap_values[0], show=False)

# Alteration 1
# plt.tight_layout()

# Alteration 2
# plt.rcParams['figure.constrained_layout.use'] = True

plt.show()
```

## No alterations

y-axis labels are cut off in both regular and fullscreen modes.

### Regular `plt.show()`
![no_tight_regular](https://github.com/shap/shap/assets/82195909/6fc23e0c-2e48-4ed1-9fb5-a5268beb6db9)

### Fullscreen
![no_tight_full_screen](https://github.com/shap/shap/assets/82195909/49c5ace9-5a83-483a-acd6-74d54041c315)

## Alteration 1: `plt.tight_layout()`

Plot is not centred in fullscreen mode - notice more whitespace on left-hand side.

### Regular `plt.show()`
![tight_regular](https://github.com/shap/shap/assets/82195909/01203392-601c-4a46-8695-26918afdd406)

### Fullscreen
![tight_full_screen](https://github.com/shap/shap/assets/82195909/cc24fd28-26af-454c-9c34-822f43658039)

## Alteration 2: Constrained (as proposed in #3200) `plt.rcParams['figure.constrained_layout.use'] = True` 

Looks good.

### Regular `plt.show()`
![constrained_regular](https://github.com/shap/shap/assets/82195909/dd6357fb-32eb-4610-a317-9fa029afaf26)

### Fullscreen
![constrained_full_screen](https://github.com/shap/shap/assets/82195909/30de07e9-3be6-4b8c-bc98-1531cbd94674)

### Minimal Reproducible Example

```python
...
```

### Expected Behavior

Alternative ways to achieve plots consistent with `plt.rcParams['figure.constrained_layout.use'] = True`.

### Bug report checklist

- [X] I have checked that this issue has not already been reported.
- [X] I have confirmed this bug exists on the [latest release](https://github.com/shap/shap/releases) of shap.
- [X] I have confirmed this bug exists on the [master branch](https://github.com/shap/shap/blob/master/CONTRIBUTING.md#installing-from-the-master-branch) of shap.
- [X] I'd be interested in making a PR to fix this bug

### Installed Versions

master

**Comment by thatlittleboy** [comment]
@dsgibbons

Can you elaborate more on what you mean by Fullscreen? Not quite sure what this means or how to replicate this.

Here is what I see:

With the code block that you provided, running in a Jupyter notebook looks fine to me (with / without any layout modifications)

![image](https://github.com/shap/shap/assets/30731072/109415eb-4706-4bb4-9e73-ccb9673a0f4e)

---

So I'm assuming the issue is in the context of using `pytest-mpl`.

```python
# tests/plots/test_issue3213.py
import pytest
import xgboost
import shap
import matplotlib.pyplot as plt

@pytest.mark.mpl_image_compare(tolerance=3)
def test_my_az():

    fig = plt.figure(layout=None)
    X, y = shap.datasets.california()
    model = xgboost.XGBRegressor().fit(X, y)
    explainer = shap.Explainer(model)
    shap_values = explainer(X)
    shap.plots.waterfall(shap_values[0], show=False)

    return fig

# pytest tests/plots/test_issue3213.py --mpl-generate-path=baseline2
```

### no layout (None)

![image](https://github.com/shap/shap/assets/30731072/ba866e3f-0aac-4e97-b23a-fe1b0352e4ee)

Labels are cut-off, as reported.

### tight layout (which is what most of our tests use)

![image](https://github.com/shap/shap/assets/30731072/da1d0543-4e24-4790-bba3-79f63ad864f5)


### constrained layout

![image](https://github.com/shap/shap/assets/30731072/6485f4ec-0f97-4ec3-950f-3759d904137c)

---

So I think I can replicate the regular `plt.show()`, but not the `fullscreen` one. Not sure what is meant there.

**Comment by thatlittleboy** [comment]
By the way, some general comments/thoughts that should be relevant to the problem at hand:

* Layout is a property of the Figure, not the Axes.
* As we plan to move to an axes-based API in 0.44 (e.g. exposing the `ax=` argument uniformly to all of our plot functions), we should be manipulating the axes object rather than making changes to the Figure in general.
* When we do so, my inclination is that when `ax=None` (not supplied by user), we will generate the Figure + Axes internally, in which case we can control the layout during instantiation, e.g. `plt.figure(layout="constrained")` or `"tight"`, depending on our conclusions here. If the user chooses to pass in an `ax`, then all bets are off and they should control the figure layout themselves.
* To summarize, I'm not convinced on the use of `plt.rcParams` in `shap` to control the look of our plots since it is a global setting. As to Contrained v.s. tight, I think we can control it locally during the instantiation of the figure whichever one we choose.
* I honestly can't remember what the issue I had with "constrained", but most likely it is one of the [limitations mentioned in the docs](https://matplotlib.org/stable/tutorials/intermediate/constrainedlayout_guide.html#limitations).

**Comment by dsgibbons** [comment]
By fullscreen, I mean clicking the maximize button of the plot. I'm not using a notebook.

The plots In this issue were generated using a Macbook Pro with a 2560 x 1600 display. I've since regenerated the plots using an Ubuntu machine with a 2560 x 1440 display. 

I can summarise the plots on my Ubuntu machine as follows:
1) No alterations looks good in both regular and fullscreen.
2) `plt.tight_layout` looks good in regular, but still off-centre in fullscreen.
3) The constrained layout looks good in both regular and fullscreen.

> To summarize, I'm not convinced on the use of plt.rcParams in shap to control the look of our plots since it is a global setting. As to Contrained v.s. tight, I think we can control it locally during the instantiation of the figure whichever one we choose.

That's fine. At the time I didn't realise that there were more localized functions such as `plt.subplots(layout="constrained")`.

> When we do so, my inclination is that when ax=None (not supplied by user), we will generate the Figure + Axes internally, in which case we can control the layout during instantiation, e.g. plt.figure(layout="constrained") or "tight", depending on our conclusions here. If the user chooses to pass in an ax, then all bets are off and they should control the figure layout themselves.

Agreed. Sounds good to me.

**Comment by thatlittleboy** [comment]
@dsgibbons Got it, let's go with `constrained_layout` in #3200 (but control via `plt.figure(layout="constrained")` rather than the global rcParams) and close this issue.

**Comment by Pratham-ja** [comment]
Hi, @CloseChoice, I’d like to work on this issue.
It seems from discussion that `constrained layout` works better for the plot. I plan to handle this at figure level only as you also suggested. Please, let me know if this looks good to open a PR?

**Comment by crocmons** [comment]
Hello @connortann  @thatlittleboy @dsgibbons, I'd like to work on this issue.

Based on the issue description, the current approaches have tradeoffs:
- `plt.tight_layout()` fixes label cutoff but causes uncentered plots in fullscreen mode
- `plt.rcParams['figure.constrained_layout.use'] = True` looks good but has global side effects

I propose using `plt.figure(layout="constrained")` as a better solution. This approach:
- Provides consistent automatic spacing without the centering issues of `tight_layout()`
- Avoids global state changes from setting rcParams
- Is the modern matplotlib-recommended approach for layout management
- Can be applied per-figure for explicit, predictable behavior

To implement this consistently across all plot types, I would:
1. Add a helper function [create_figure()](cci:1://file:///d:/github-projects/shap/shap/plots/_utils.py:277:0-306:18) in [_utils.py](cci:7://file:///d:/github-projects/shap/shap/plots/_utils.py:0:0-0:0) that creates figures with constrained layout
2. Update all plot functions to use this helper when creating new figures
3. Ensure plots that rely on implicit figure creation (like waterfall) explicitly create figures first

This would provide a consistent, robust solution for all SHAP plot types. 
Soon I'll raise a PR for this; let me know. Would this approach sound good to you?
